lib/elixir/lib/regex.ex

```diff
@@ -647,13 +647,12 @@ defmodule Regex do
 
   def replace(regex, string, replacement, options)
       when is_binary(string) and is_binary(replacement) and is_list(options) do
-    do_replace(regex, string, precompile_replacement(replacement), options)
+    do_replace(regex, string, replacement, options)
   end
 
   def replace(regex, string, replacement, options)
       when is_binary(string) and is_function(replacement) and is_list(options) do
-    {:arity, arity} = Function.info(replacement, :arity)
-    do_replace(regex, string, {replacement, arity}, options)
+    do_replace(regex, string, replacement, options)
   end
```

Comment:
The `is_function/1` guard is now irrelevant since the match is on `precompile_replacement/1`. 

We could simplify the code by removing the guard and the `do_replace/4` private function at the same time…

```elixir
  @spec replace(t, String.t(), String.t() | (... -> String.t()), [term]) :: String.t()
  def replace(regex, string, replacement, options \\ []) do
    opts = if Keyword.get(options, :global) != false, do: [:global], else: []
    opts = [{:capture, :all, :index} | opts]
    case safe_run(regex, string, opts) do
      :nomatch ->
        string

      {:match, [mlist | t]} when is_list(mlist) ->
        apply_list(string, replacement, [mlist | t]) |> IO.iodata_to_binary()
        apply_list(string, precompile_replacement(replacement), [mlist | t])
        |> IO.iodata_to_binary()

      {:match, slist} ->
        apply_list(string, replacement, [slist]) |> IO.iodata_to_binary()
        apply_list(string, precompile_replacement(replacement), [slist]) |> IO.iodata_to_binary()
    end
  end
```

Reply:
Yes, that's a good refactoring.
But since this PR is already merged, maybe you can open a PR for this change?
I don't want to steal your credit 😉